draft-ietf-quic-load-balancers.md

```diff
@@ -444,71 +444,157 @@ routing information of any given 4-tuple.
-this padded server-id intermediate using its key to generate a mask which it
-applies to the encrypted nonce. This provides the decrypted nonce value.
+~~~pseudocode
+expand(0xaaba3c, 0x13) = 0xaaba3c00000000000000000000000013
+~~~
 
-nonce = encrypted_nonce ^ AES-ECB(key, padded-server_id_intermediate)
+The example at the end of this section helps to clarify the steps described
+below.
 
-* Pass 3:  The load balancer decrypts the server ID using 128-bit AES ECB mode.
-The nonce octets are zero-padded to 16 octets.  AES-ECB encrypts this nonce
-using its key to generate a mask which it applies to the intermediate server id.
-This provides the decrypted server ID.
+1. The server concatenates the server ID and nonce to create plaintext_CID.
 
-server_id = server_id_intermediate ^ AES-ECB(key, padded-nonce)
+2. The server splits plaintext_CID into components left_0 and right_0 of equal
+length, splitting an odd octet in half if necessary.
 
-For example, if the nonce length is 10 octets and the server ID length is 2
-octets, the connection ID can be as small as 13 octets.  The load balancer uses
-the the second through eleventh octets of the connection ID for the nonce,
-zero-pads it to 16 octets, uses xors the result with the twelfth and thirteenth
-octet. The result is padded with 14 octets of zeros and encrypted to obtain a
-mask that is xored with the nonce octets. Finally, the nonce octets are padded
-with six octets of zeros, encrypted, and the first two octets xored with the
-server ID octets to obtain the actual server ID.
+3. Encrypt left_0. The encryption is 128-bit AES-ECB with the key provided by
+the configuration agent, and the plaintext argument is an expanded version of
+left_0 where left_0 constitutes the most significant bits, 0x01 is the least
+significant octet, and all other bits are zero.
 
-This three-pass algorithm is a simplified version of the FFX algorithm, with
-the property that each encrypted nonce value depends on all server ID bits, and
-each encrypted server ID bit depends on all nonce bits and all server ID bits.
-This mitigates attacks against stream ciphers in which attackers simply flip
-encrypted server-ID bits.
+4. XOR the most significant bits of the ciphertext with right_0 to form
+right_1.
 
-The output of the decryption is the server ID that the load balancer uses for
-routing.
+    Thus steps 3 and 4 can be expressed as
 
-### Server Actions
+    ```
+    right_1 = right_0 ^ truncate(AES_ECB(key, expand(left_0, 0x01))
+    ```
 
-When generating a routable connection ID, the server writes arbitrary bits into
-its nonce octets, and its provided server ID into the server ID octets. See
-{{cid-entropy}} for nonce generation considerations.
+5. Repeat steps 3 and 4, but use them to compute left_1 by expanding and
+encrypting right_1 with the least significant octet as 0x02 and XOR the
+results with left_0.
+
+    ```
+    left_1 = left_0 ^ truncate(AES_ECB(key, expand(right_1), 0x02))
+    ```
+
```

Comment:
The truncate operation here is underspecified. Left might not start on an even octet boundary. If it does not, the top 4 bits of the AES_ECB result should be reset to zero before the XOR.

Reply:
I chose to bitshift right so that the MSB is on the octet boundary